include/convex_bodies/ellipsoid.h

```diff
@@ -0,0 +1,288 @@
+//Contributed and/or modified by Vaibhav Thakkar, as part of Google Summer of Code 2021 program.
+
+// Licensed under GNU LGPL.3, see LICENCE file
+
+
+#ifndef ELLIPSOIDS_H
+#define ELLIPSOIDS_H
+
+#include <iostream>
+#include <Eigen/Eigen>
+#include <boost/math/special_functions/gamma.hpp>
+
+
+template <typename NT>
+NT log_gamma_function(NT x)
+{
+    if (x <= NT(100)) return std::log(tgamma(x));
+    return (std::log(x - NT(1)) + log_gamma_function(x - NT(1)));
+}
+
+
+template <class Point, class MT>
+class Ellipsoid{
+private:
+    typedef typename Point::FT NT;
+    typedef typename std::vector<NT>::iterator viterator;
+    typedef Eigen::Matrix<NT, Eigen::Dynamic, 1> VT;
+
+    // representation is (x - c)' A (x - c) <= 1, center is assumed to be origin for now
+    MT A;
+    Point c;
+
+    // MT L;   // LL' = A
+    unsigned int dim;
+
+    // eigen vectors and values
+    VT eigen_values;
+    VT eigen_values_inv;
+    VT eigen_values_inv_sqrt;
+    MT eigen_vecs;
+
+public:
+
+    Ellipsoid(MT& Ain) : A(Ain) {
+        // Eigen::LLT<Eigen::MatrixXd> lltOfA(A); // compute the Cholesky decomposition of Ain
+        // if(lltOfA.info() == Eigen::NumericalIssue) {
+        //     throw std::runtime_error("Possibly non semi-positive definitie matrix!");
+        // }
+        // L = lltOfA.matrixL();
+
+        Eigen::SelfAdjointEigenSolver<MT> eigensolver(A);
+        if (eigensolver.info() != Eigen::Success) {
+            throw std::runtime_error("Eigen solver returned error!");
+        }
+
+        eigen_values = eigensolver.eigenvalues();
+        eigen_vecs = eigensolver.eigenvectors();
+
+        eigen_values_inv = eigen_values.array().inverse().matrix();
+        eigen_values_inv_sqrt = eigen_values_inv.array().sqrt().matrix();
+
+        dim = A.rows();
+        c = Point(dim);
+    }
+
+
+    // Constructor for copula ellipsoid
+    Ellipsoid(std::vector<std::vector<NT> >& Ain) {
+        dim = Ain.size();
+        A.resize(dim, dim);
+        for (unsigned int i = 0; i < Ain.size(); i++) {
+            for (unsigned int j = 0; j < Ain.size(); j++) {
+                A(i,j) = Ain[i][j];
+            }
+        }
+
+        Eigen::SelfAdjointEigenSolver<MT> eigensolver(A);
+        if (eigensolver.info() != Eigen::Success) {
+            throw std::runtime_error("Eigen solver returned error!");
+        }
+
+        eigen_values = eigensolver.eigenvalues();
+        eigen_vecs = eigensolver.eigenvectors();
+
+        eigen_values_inv = eigen_values.array().inverse().matrix();
+        eigen_values_inv_sqrt = eigen_values_inv.array().sqrt().matrix();
+
+        dim = A.rows();
+        c = Point(dim);
+    }
+
+
+    VT eigenvals() const {
+        return eigen_values;
+    }
+
+
+    VT eigenvals_inv() const {
+        return eigen_values_inv;
+    }
+
+
+    VT eigenvals_inv_sqrt() const {
+        return eigen_values_inv_sqrt;
+    }
+
+
+    MT eigenvecs() const {
```

Comment:
why not `eigenvectors`?

Reply:
did this to avoid confusion between Eigen's `eigenvectors` function and our custom function. 
Want to know your suggestion on this, should I change it to the `eigenvectors` ?

Reply:
As above

Reply:
renamed to `eigenvectors` 👍🏽